internal/webhook/external/providers/github/provider.go

```diff
@@ -0,0 +1,102 @@
+package github
+
+import (
+	"errors"
+	"io"
+	"net/http"
+	"os"
+
+	gh "github.com/google/go-github/v71/github"
+
+	xhttp "github.com/akuity/kargo/internal/http"
+)
+
+var (
+	// ErrMissingSignature is returned when the 'X-Hub-Signature-256'
+	// header is not found or empty.
+	ErrMissingSignature = errors.New("missing signature")
+	// ErrInvalidSignature is returned when the 'X-Hub-Signature-256'
+	// header is not the value that was expected.
+	ErrInvalidSignature = errors.New("invalid signature")
+	// ErrSecretUnset is returned when the 'GH_WEBHOOK_SECRET'
+	// environment variable is empty.
+	ErrSecretUnset = errors.New("secret is unset")
+	// ErrUnsupportedEventType is returned when the received
+	// request body does not conform to a github push event.
+	ErrUnsupportedEventType = errors.New("unsupported event type")
+)
+
+type provider struct {
+	secret string
+}
+
+func NewProvider() (*provider, error) {
+	secret, ok := os.LookupEnv("GH_WEBHOOK_SECRET")
+	if !ok {
+		return nil, ErrSecretUnset
+	}
+	return &provider{secret: secret}, nil
+}
+
+func (p *provider) Name() string {
+	return "github"
+}
+
+func (p *provider) GetRepository(r *http.Request) (string, error) {
+	var signature string
+	if h := r.Header.Get(gh.SHA1SignatureHeader); h != "" {
+		signature = h
+	}
+	if h := r.Header.Get(gh.SHA256SignatureHeader); h != "" {
+		signature = h
+	}
+	if signature == "" {
+		return "", xhttp.UnauthorizedError(ErrMissingSignature)
```

Comment:
Nit: It might be more accurate to classify this as _unauthenticated_ (which would be a 401 instead of a 403). The nuanced difference is not allowed because you don't have permissions (403) vs not allowed because you didn't identify yourself (401).

If you wanted to stick with unauthorized / 403 I'm not bothered by it.

Dealer's choice on whether to change or keep as is.

Reply:
Arguably a webhook missing a required header could also be considered a bad request.

However you want to deal with this is fine.